### ISSUE TITLE
RFC: Typography

### ISSUE THREAD

**Issue by Filippinifra** [casavo/worms-ui#1]
Component: Typography
[Figma component](https://www.figma.com/file/0vYcLbHGHFIJ44hFI45WQJ/%F0%9F%9A%80-BASE-UI?node-id=0%3A1&t=c0Gqv3sAu0IgYzTc-0)

Props:
`  variant: 'display' | 'h1' | 'h2' | 'h3' | 'h4' | 'description' | 'bodyM' | 'bodyL' | 'caption' | 'smallText'`
`  children: React.ReactNode`
`  weight?: '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900'`
`  color?: string`
`  style?: CSSProperties`
`  dataTestId?: string`

I suggest handling the mobile logic inside the style of the component: for example, the h3 variant has a different font size on mobile and on desktop, so I want to handle it using something like that (emotion variant implementation)

```
...
h3: [
    base,
    interFontStyle,
    {
      fontSize: 28,
      fontWeight: 600,
      lineHeight: '120%',
      '@media': {
        [mediaQueryDimension.smallAndMedium]: {
          fontSize: 24,
        },
      },
    },
  ],
...
```

Then, as you can see from figma's mock, there are other variants called "underline", "regular", "strong", I think we can model this behaviour using other props (and not the variant one) such as the "style" prop in order to pass additional style (for example the underline text decoration), or to use the "weight" prop to pass regular or strong font-weight

------------------------------------

This is the implementation that I did for efesto:

index.tsx
```typescript
import { CSSProperties, FC } from 'react'
import { typographyStyles } from './styles.css'

export type TypographyVariants = 'h2' | 'h3' | 'description' | 'bodyM' | 'bodyL' | 'caption'

export interface TypographyProps {
  variant: TypographyVariants
  children: React.ReactNode
  weight?: '100' | '200' | '300' | '400' | '500' | '600' | '700' | '800' | '900'
  color?: string
  style?: CSSProperties
  dataTestId?: string
}

export const Typography: FC<TypographyProps> = ({ variant, children, weight, color, style, dataTestId }) => {
  const { tag, classname } = styles[variant]
  const Tag = tag as React.ElementType
  return (
    <Tag className={classname} style={{ fontWeight: weight, color, ...style }} data-testid={dataTestId}>
      {children}
    </Tag>
  )
}

const styles = {
  h2: {
    tag: 'h2',
    classname: typographyStyles.h2,
  },
  h3: {
    tag: 'h3',
    classname: typographyStyles.h3,
  },
  description: {
    tag: 'p',
    classname: typographyStyles.description,
  },
  bodyM: {
    tag: 'p',
    classname: typographyStyles.bodyM,
  },
  bodyL: {
    tag: 'p',
    classname: typographyStyles.bodyL,
  },
  caption: {
    tag: 'p',
    classname: typographyStyles.caption,
  },
}

```

style.css.ts
```typescript
import { mediaQueryDimension } from '@app/utils/breakpoints'
import { colors } from '@casavo/base-ui'
import { style, styleVariants } from '@vanilla-extract/css'

const base = style({ color: colors.greyscale600, overflowWrap: 'anywhere' })
const interFontStyle = style({ fontFamily: 'Inter' })
const bagossFontStyle = style({ fontFamily: 'Bagoss Condensed' })

export const typographyStyles = styleVariants({
  h2: [
    base,
    bagossFontStyle,
    {
      fontSize: 32,
      fontWeight: 300,
      lineHeight: '120%',
      '@media': {
        [mediaQueryDimension.smallAndMedium]: {
          fontSize: 28,
        },
      },
    },
  ],
  h3: [
    base,
    interFontStyle,
    {
      fontSize: 28,
      fontWeight: 600,
      lineHeight: '120%',
      '@media': {
        [mediaQueryDimension.smallAndMedium]: {
          fontSize: 24,
        },
      },
    },
  ],
  description: [
    base,
    interFontStyle,
    {
      fontSize: 14,
      fontWeight: 400,
      lineHeight: '150%',
    },
  ],
  bodyM: [
    base,
    interFontStyle,
    {
      fontSize: 16,
      fontWeight: 400,
      lineHeight: '150%',
    },
  ],
  bodyL: [
    base,
    interFontStyle,
    {
      fontSize: 20,
      fontWeight: 400,
      lineHeight: '150%',
      '@media': {
        [mediaQueryDimension.smallAndMedium]: {
          fontSize: 18,
        },
      },
    },
  ],
  caption: [
    base,
    interFontStyle,
    {
      fontSize: 12,
      fontWeight: 400,
      lineHeight: '150%',
    },
  ],
})

```

This implementation handles only a few variants because in Efesto we don't use the entire set of variants: when I will create the pr for the creation of the component I will add all the variants

**Comment by simonedavico** [comment]
While reasoning about this, I realized we are probably missing some basics:

* Atomic, reusable styles
* A theme definition
* A way to extend-customize it for apps that may need it

I would solve these basic problems before implementing any component, to ensure that we have in place a good, extendable system to style our components. Since our CSS stack includes `vanilla-extract`, I started looking for approaches to do this. I found [Bento Design System](https://bento.buildo.io/), a DS developed by a software house in Milan that uses the same tech stack. It is open source, and I think their approach is very nice:

* They setup a type-safe theme shape
* The theme itself is implemented under the hood by CSS vars, so it's overridable without any JS for projects that do not use any special library
* The above take very little code, since they leverage functionalities of `sprinkles`

The first thing I would do is look more into this approach and define a theme shape based on Casavo Base UI atoms. After we have a good system in place, we will use it to build components. What do you think?

**Comment by Filippinifra** [comment]
I agree with all this stuff! The purpose (as you said to me in private chat) could be to create something that is reusable for worms, efesto and the other teams because components are very similar and the only thing that change is the theme, so in this way, we can structure our theme and then allow other projects to set their with the components

**Comment by simonedavico** [comment]
Implemented in #10
